File: Python/Problem 021.py
import time

import math

start = time.time()


def d(n):
    total = 1
    for x in range(2, math.ceil(math.sqrt(n))):
        if n % x == 0:
            total += x
            if (n / x) != x:
                total += (n / x)
    return total


amicable_nums = set()

for a in range(1, 10000):
    if a not in amicable_nums:
        b = d(a)
        if d(a) == b and d(b) == a and a != b:
            amicable_nums.add(a)
            amicable_nums.add(b)

sum = 0
for x in amicable_nums:
    sum += x
print(sum)
print(amicable_nums)

print(time.time() - start)

File: Java/Problem_021.java
package ProjectEuler;

import java.util.ArrayList;
import java.util.List;

public class Problem_021 {

    public static void main(String[] args) {
        double start = System.nanoTime();

        List<Integer> amicable_nums = new ArrayList<>();
        for (int a = 1; a < 10000; a++) {
            if (!checkPresent(a, amicable_nums)) {
                int b = d(a);
                if (a != b && d(b) == a) {
                    amicable_nums.add(a);
                    amicable_nums.add(b);
                }
            }
        }

        long sum = 0;
        for (int n : amicable_nums) {
            sum += n;
        }

        System.out.println(sum);
        System.out.println(amicable_nums);

        System.out.println((System.nanoTime() - start) / 1_000_000 + " ms");
    }

    private static int d(int n) {
        if (n == 1) {
            return 1;
        }
        int total = 1;
        for (double i = 2; i <= Math.sqrt(n); i++) {
            if (n % i == 0) {
                total += i + (n / i);
            }
        }
        if (Math.sqrt(n) == (int) Math.sqrt(n)) {
            total -= Math.sqrt(n);
        }
        return total;
    }

    private static boolean checkPresent(int n, List<Integer> amics) {
        for (int i : amics) {
            if (i == n) {
                return true;
            }
        }
        return false;
    }
}